staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -133,9 +133,18 @@ func (s sortableWatchCacheEvents) Swap(i, j int) {
 // returned by Next() need to be events from a List() done on the underlying store of
 // the watch cache.
 // The items returned in the interval will be sorted by Key.
-func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc) (*watchCacheInterval, error) {
+func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc, key string) (*watchCacheInterval, error) {
 	buffer := &watchCacheIntervalBuffer{}
-	allItems := store.List()
```

Comment:
@wojtek-t is the result returned by `store.List()` immutable? If so, maybe we can delay the buffer calculation to buffer.Next()

Reply:
I don't understand the question, 
The store changes over time, so we can't really delay it (and it won't buy us anything).

Reply:
I think the reason it has relatively higher memory usage is the following code:

```
  149         for i, item := range allItems {
  150                 elem, ok := item.(*storeElement)
  151                 if !ok {
  152                         return nil, fmt.Errorf("not a storeElement: %v", elem)
  153                 }
  154                 objLabels, objFields, err := getAttrsFunc(elem.Object)
  155                 if err != nil {
  156                         return nil, err
  157                 }
  158                 buffer.buffer[i] = &watchCacheEvent{. # likely this line
  159                         Type:            watch.Added,
  160                         Object:          elem.Object,
  161                         ObjLabels:       objLabels,
  162                         ObjFields:       objFields,
  163                         Key:             elem.Key,
  164                         ResourceVersion: resourceVersion,
  165                 }
  166                 buffer.endIndex++
  167         }
```

so if `allItems` returned by `store.List()` is immutable, we can retain it in the `watchCacheInterval` (sorting required), and in `watchCacheInterval.Next()` we populate one item to minimize the memory usage. By doing this, the lock period is also decreased.

Reply:
no - the contents of store are changing over time - this won't work